File: src/main/java/com/pedigree/model/Individual.java
package com.pedigree.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class Individual {
    private final String id;
    private String firstName;
    private String lastName;
    private Gender gender;
    // Dates are stored as free-form phrases (e.g., "ABT 1900", "BET 1890 AND 1895", "6 MAY 1981", etc.)
    private String birthDate;
    private String birthPlace;
    private String deathDate;
    private String deathPlace;
    private final List<Note> notes;
    private final List<MediaAttachment> media;
    private final List<Tag> tags;

    public Individual(String firstName, String lastName, Gender gender) {
        this(UUID.randomUUID().toString(), firstName, lastName, gender);
    }

    public Individual(String id, String firstName, String lastName, Gender gender) {
        this.id = Objects.requireNonNull(id, "id");
        this.firstName = Objects.requireNonNull(firstName, "firstName");
        this.lastName = Objects.requireNonNull(lastName, "lastName");
        this.gender = Objects.requireNonNull(gender, "gender");
        this.notes = new com.pedigree.util.DirtyObservableList<>();
        this.media = new com.pedigree.util.DirtyObservableList<>();
        this.tags = new com.pedigree.util.DirtyObservableList<>();
    }

    @JsonCreator
    public Individual(
            @JsonProperty("id") String id,
            @JsonProperty("firstName") String firstName,
            @JsonProperty("lastName") String lastName,
            @JsonProperty("gender") Gender gender,
            @JsonProperty("birthDate") String birthDate,
            @JsonProperty("birthPlace") String birthPlace,
            @JsonProperty("deathDate") String deathDate,
            @JsonProperty("deathPlace") String deathPlace,
            @JsonProperty("notes") List<Note> notes,
            @JsonProperty("media") List<MediaAttachment> media,
            @JsonProperty("tags") List<Tag> tags
    ) {
        this.id = Objects.requireNonNull(id, "id");
        this.firstName = Objects.requireNonNull(firstName, "firstName");
        this.lastName = Objects.requireNonNull(lastName, "lastName");
        this.gender = Objects.requireNonNull(gender, "gender");
        this.birthDate = birthDate;
        this.birthPlace = birthPlace;
        this.deathDate = deathDate;
        this.deathPlace = deathPlace;
        this.notes = (notes != null) ? new com.pedigree.util.DirtyObservableList<>(notes) : new com.pedigree.util.DirtyObservableList<>();
        this.media = (media != null) ? new com.pedigree.util.DirtyObservableList<>(media) : new com.pedigree.util.DirtyObservableList<>();
        this.tags = (tags != null) ? new com.pedigree.util.DirtyObservableList<>(tags) : new com.pedigree.util.DirtyObservableList<>();
    }

    public String getId() { return id; }
    public String getFirstName() { return firstName; }
    public void setFirstName(String firstName) { this.firstName = firstName; com.pedigree.util.DirtyFlag.setModified(); }
    public String getLastName() { return lastName; }
    public void setLastName(String lastName) { this.lastName = lastName; com.pedigree.util.DirtyFlag.setModified(); }
    public Gender getGender() { return gender; }
    public void setGender(Gender gender) { this.gender = gender; com.pedigree.util.DirtyFlag.setModified(); }
    public String getBirthDate() { return birthDate; }
    public void setBirthDate(String birthDate) { this.birthDate = birthDate; com.pedigree.util.DirtyFlag.setModified(); }
    public String getBirthPlace() { return birthPlace; }
    public void setBirthPlace(String birthPlace) { this.birthPlace = birthPlace; com.pedigree.util.DirtyFlag.setModified(); }
    public String getDeathDate() { return deathDate; }
    public void setDeathDate(String deathDate) { this.deathDate = deathDate; com.pedigree.util.DirtyFlag.setModified(); }
    public String getDeathPlace() { return deathPlace; }
    public void setDeathPlace(String deathPlace) { this.deathPlace = deathPlace; com.pedigree.util.DirtyFlag.setModified(); }
    public List<Note> getNotes() { return notes; }
    public List<MediaAttachment> getMedia() { return media; }
    public List<Tag> getTags() { return tags; }
}

File: src/main/kotlin/com/family/tree/model/Individual.kt
package com.family.tree.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.family.tree.util.DirtyFlag
import com.family.tree.util.DirtyObservableList
import java.util.Objects
import java.util.UUID

@JsonIgnoreProperties(ignoreUnknown = true)
class Individual {
    val id: String
    var firstName: String? = null
        set(value) { field = value; DirtyFlag.setModified() }
    var lastName: String? = null
        set(value) { field = value; DirtyFlag.setModified() }
    var gender: Gender? = null
        set(value) { field = value; DirtyFlag.setModified() }
    val events: MutableList<GedcomEvent>
    val sources: MutableList<SourceCitation>
    val notes: MutableList<Note>
    val media: MutableList<MediaAttachment>
    val tags: MutableList<Tag>

    constructor(firstName: String, lastName: String, gender: Gender) : this(
        UUID.randomUUID().toString(), firstName, lastName, gender
    )

    constructor(id: String, firstName: String, lastName: String, gender: Gender) {
        this.id = Objects.requireNonNull(id, "id")
        this.firstName = Objects.requireNonNull(firstName, "firstName")
        this.lastName = Objects.requireNonNull(lastName, "lastName")
        this.gender = Objects.requireNonNull(gender, "gender")
        this.events = DirtyObservableList<GedcomEvent>()
        this.sources = DirtyObservableList<SourceCitation>()
        this.notes = DirtyObservableList<Note>()
        this.media = DirtyObservableList<MediaAttachment>()
        this.tags = DirtyObservableList<Tag>()
    }

    @JsonCreator
    constructor(
        @JsonProperty("id") id: String?,
        @JsonProperty("firstName") firstName: String?,
        @JsonProperty("lastName") lastName: String?,
        @JsonProperty("gender") gender: Gender?,
        @JsonProperty("events") events: List<GedcomEvent>?,
        @JsonProperty("sources") sources: List<SourceCitation>?,
        @JsonProperty("notes") notes: List<Note>?,
        @JsonProperty("media") media: List<MediaAttachment>?,
        @JsonProperty("tags") tags: List<Tag>?
    ) {
        this.id = id ?: throw NullPointerException("id")
        this.firstName = Objects.requireNonNull(firstName, "firstName")
        this.lastName = Objects.requireNonNull(lastName, "lastName")
        this.gender = Objects.requireNonNull(gender, "gender")
        this.events = if (events != null) DirtyObservableList<GedcomEvent>(events) else DirtyObservableList()
        this.sources = if (sources != null) DirtyObservableList<SourceCitation>(sources) else DirtyObservableList()
        this.notes = if (notes != null) DirtyObservableList<Note>(notes) else DirtyObservableList()
        this.media = if (media != null) DirtyObservableList<MediaAttachment>(media) else DirtyObservableList()
        this.tags = if (tags != null) DirtyObservableList<Tag>(tags) else DirtyObservableList()
    }
}
